linc_convert/utils/io/zarr/drivers/zarr_python.py

```diff
@@ -0,0 +1,413 @@
+                fill_value=getattr(base_level._array, "fill_value", None),
+                order=getattr(base_level._array, "order", None),
+                attributes=getattr(
+                        getattr(base_level._array, "metadata", None), "attributes", None
+                        ),
+                overwrite=True,
+                )
+        # Handle extra options based on metadata type
+        meta = getattr(base_level, "metadata", None)
+        if meta is not None:
+            if hasattr(meta, "dimension_separator"):
+                opts["chunk_key_encoding"] = _dimension_separator_to_chunk_key_encoding(
+                        meta.dimension_separator, 2
+                        )
+            if hasattr(meta, "chunk_key_encoding"):
+                opts["chunk_key_encoding"] = getattr(meta, "chunk_key_encoding", None)
+            if hasattr(base_level, "serializer"):
+                opts["serializer"] = getattr(base_level, "serializer", None)
+            if hasattr(meta, "dimension_names"):
+                opts["dimension_names"] = getattr(meta, "dimension_names", None)
+        # Remove None values
+        opts = {k: v for k, v in opts.items() if v is not None}
+        opts.update(kwargs)
+        arr = self._zgroup.create_array(name=name, shape=shape, **opts)
+        if data is not None:
+            arr[:] = data
+        return ZarrPythonArray(arr)
+
+
+def _make_compressor(
+        name: str | None, zarr_version: Literal[2, 3], **prm: dict
+        ) -> CompressorsLike:
+    """Build compressor object from name and options."""
+    if not isinstance(name, str):
+        return name
+
+    if zarr_version == 2:
+        import numcodecs
+
+        compressor_map = {
+            "blosc": numcodecs.Blosc,
+            "zlib": numcodecs.Zstd,
+            }
+    elif zarr_version == 3:
+        import zarr.codecs
+
+        compressor_map = {
+            "blosc": zarr.codecs.BloscCodec,
+            "zlib": zarr.codecs.ZstdCodec,
+            }
+    else:
+        raise ValueError()
+    name = name.lower()
+
+    if name not in compressor_map:
+        raise ValueError("Unknown compressor", name)
+    Compressor = compressor_map[name]
+
+    return Compressor(**prm)
+
+
+def _dimension_separator_to_chunk_key_encoding(
+        dimension_separator: Literal[".", "/"], zarr_version: Literal[2, 3]
+        ) -> ChunkKeyEncodingLike:
+    dimension_separator = dimension_separator
+    if dimension_separator == "." and zarr_version == 2:
+        pass
+    elif dimension_separator == "/" and zarr_version == 3:
+        pass
+    else:
+        dimension_separator = ChunkKeyEncodingParams(
+                name="default" if zarr_version == 3 else "v2",
+                separator=dimension_separator
+                )
+        return dimension_separator
+
+
+SHARD_FILE_SIZE_LIMIT = (
+        2  # compression ratio
+        * 2  # TB
+        * 2 ** 30  # GB->Bytes
+)
```

Comment:
This was when we were assuming that memory-efficient shard writing was possible (at least when writing from scratch), but it now seems that neither zarr-python not tensorstore do that efficiently (?). So should we use a more conservative shard limit? Or maybe have "max memory use" a config option?